### PR TITLE
#164829209 implement commenting on articles feature

### PR DESCRIPTION
**Pull request by bl4ck4ndbr0wn**
#### Description
As a User:
I want to be able to make a comment(s) on articles, both authored by me and those not authored by me.

#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [x] End to end
- [x] Integration

#### Checklist:
- [x] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [ ]  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### How should this be manually tested?
1.  Clone the repository 
```https://github.com/andela/ah-centauri-backend.git```
2. On completing of clonning
```bash 
cd ah-centauti-backend
git checkout feature/164829209-user-can-comment-on-article
```
3. Rename the .env.example file into .env and update the parameters with your own keys.

> For Centauri team members I will share this keys through slack

4. Start development server and send the following through postman

**Displaying all comments for an Article**
```
GET /api/articles/:slug/comments
returns All Comment. 
```
**Displaying all comments replies for an Article**
```
GET /api/articles/:slug/comments?parent=9
returns All Comment. 
```
**Add Comments to an Article**
```
POST /api/articles/:slug/comments

Example request body:

{
  "comment": {
    "body": "His name was my name too."
  }
}
Authentication required returns the created Comment. Required field: body
```
**Add a Reply to a Comment in an Article**
```
POST /api/articles/:slug/comments

Example request body:

{
  "comment": {
    "body": "His name was my name too.",
    "parent": 9
  }
}

args: 
       :parent: is the ID of the comment you are repling to.

Authentication required returns the created Comment. Required field: body
```
**Update a Comment to an Article**
```
PATCH /api/articles/:slug/comments/:id

Example request body:

{
  "comment": {
    "body": "His name was my name too."
  }
}
Authentication required, returns the created Comment. Required field: body
```
**Delete Comment**
```
DELETE /api/articles/:slug/comments/:id

Authentication required.
```

#### Pivotal Tracker
[#164829209](https://www.pivotaltracker.com/n/projects/2321088/stories/164829209)

### CODE REVIEW

**Comment by Sm0k3s** [comment]
Looks pragmatic to me

**Review by Sm0k3s — APPROVED** [review]

**Review by ValerieRono — CHANGES_REQUESTED** [review]
1). change urls to follow conventions
2). replies to a comment's comment not reflecting

**Review by ValerieRono — APPROVED** [review]

File: authors/apps/comments/tests/test_comment_view.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+
+def specific_articles_url(slug):
+    return reverse("comments:comments", args=[slug])
+
+
+def specific_comment_url(slug, comment_id):
+    return reverse('comments:list_comment', args=[slug, comment_id])
+
+
```
Comment (oagutu) [inline authors/apps/comments/tests/test_comment_view.py:48]: Why not have a `BaseTestCase` with a "parent"  `setUp` method. The Test classes in this file can then simply inherit form the `BaseTestCase`?

File: authors/apps/comments/views.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            return Response({"errors": "Specified Article does not exist."}, status=status.HTTP_404_NOT_FOUND)
+        try:
+            comment = Comment.objects.get(pk=self.kwargs['id'])
+            serializer = self.serializer_class(comment)
+            return Response(serializer.data,
+                            status=status.HTTP_200_OK)
+        except Comment.DoesNotExist:
+            return Response({"errors": "Specified Comment does not exist."}, status=status.HTTP_404_NOT_FOUND)
+
+    def post(self, request, *args, **kwargs):
```
Comment (oagutu) [inline authors/apps/comments/views.py]: Can't this already be handled by the `post` method in the class ` RetrieveCommentAPIView` above?

File: authors/apps/comments/views.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        try:
+            article = Articles.objects.get(slug=self.kwargs['slug'])
+        except Articles.DoesNotExist:
+            return Response({"errors": "Specified Article does not exist."}, status=status.HTTP_404_NOT_FOUND)
+        try:
+            comment = Comment.objects.get(pk=self.kwargs['id'])
+            self.check_object_permissions(request, comment)
+            comment.delete()
+            return Response({'message': "Deletion successful."}, status.HTTP_204_NO_CONTENT)
```
Comment (oagutu) [inline authors/apps/comments/views.py]: `204` responses should not contain a message body

**Review by oagutu — COMMENTED** [review]
@bl4ck4ndbr0wn good job with this. Look into the issues raised, including the following:
Improve test coverage
`GET` - `/api/articles/<slug>/comments/`: 
- Repetition of comments in response ie. A child is returned under a parent as well as on its own

File: authors/apps/comments/views.py (hunk 2 above)
Comment (bl4ck4ndbr0wn) [reply authors/apps/comments/views.py]: The post in:
RetrieveCommentAPIView requires one argument which is the slug to create a new comment.
RetrieveUpdateDeleteCommentAPIView requires two arguments which are the slug and comment id to create a new comment[reply]

Comment (bl4ck4ndbr0wn) [reply authors/apps/comments/views.py]: Each view has it own url that limits what can be sent by the user.

File: authors/apps/comments/views.py (hunk 3 above)
Comment (bl4ck4ndbr0wn) [reply authors/apps/comments/views.py]: > @bl4ck4ndbr0wn good job with this. Look into the issues raised, including the following:
> Improve test coverage
> `GET` - `/api/articles/<slug>/comments/`:
> 
> * Repetition of comments in response ie. A child is returned under a parent as well as on its own

In regard to test coverage currently, I have tested everything inside the comments application.
I have uploaded an image showing the coverage in my feature.
image

Maybe to add the coverage if it is ok, i can test other apps in the project folder.

File: authors/apps/comments/views.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        try:
+            article = Articles.objects.get(slug=slug)
+            comments = Comment.objects.all().filter(article_id=article.id)
+            serializer = self.serializer_class(comments, many=True)
+
+            return Response(serializer.data,
+                            status=status.HTTP_200_OK)
+        except Articles.DoesNotExist:
+            return Response({"errors": "Specified Article does not exist."}, status=status.HTTP_404_NOT_FOUND)
```
Comment (ogiste) [inline authors/apps/comments/views.py]: Create a separate file for the response messages sent to in a response, makes it easier to update.

**Pull request merged**